cacheops/query.py

```diff
@@ -86,11 +88,23 @@ def wrapper(*args, **kwargs):
             cache_key = 'as:' + _get_key(func, args, kwargs, key_extra)
 
             cache_data = redis_client.get(cache_key)
+
             if cache_data is not None:
                 return pickle.loads(cache_data)
 
-            result = func(*args)
+            if USE_LOCK:
+                with redis_lock.Lock(redis_client, cache_key):
+                    cache_data = redis_client.get(cache_key)
+
+                    if cache_data is not None:
+                        return pickle.loads(cache_data)
+
+                    result = func(*args)
+            else:
+                result = func(*args)
+
             cache_thing(cache_key, result, cond_dnfs, timeout)
```

Comment:
Cache write should be in the lock or second process acquiring it will redo the calculation.